# Tasca M3 T02
## Exercicis amb Dataframes

### Exercici 1
Descarrega el data set Airlines Delay: Airline on-time statistics and delay causes i carrega’l a un Pandas Dataframe. Explora les dades que conté, explica breument quines variables hi ha i queda’t únicament amb les columnes que consideris rellevants. Justifica la teva elecció.

->Airlines Delay: Airline on-time statistics and delay causes

 Important
Redueix la dimensió del dataset de manera aleatòria per tal d'obtenir un dataset de només 200.000 registres. Tots els exercicis s'han de fer amb aquest dataset reduït.

- Exercici 2
Fes un informe complet del dataset:

Resumeix estadísticament el dataset i les columnes d’interès. Fes una anàlisi estadístic del que consideris rellevant.
Troba quantes dades faltants hi ha per columna.
Crea columnes noves (velocitat mitjana del vol, si ha arribat tard o no...).
Fes una taula de les aerolínies amb més endarreriments acumulats.
Quins són els vols més llargs? I els més endarrerits? Busca les rutes més llargues i les que acumulen més retards.
Aporta allò que consideris rellevant.


- Exercici 3
Exporta el dataset net i amb les noves columnes a Excel.

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import statistics
import math
import scipy.stats
%matplotlib inline

In [2]:
cd /Users/thaisrocafull/Documents/CURSOS_I_APUNTS/data_science/SPRINT 3

/Users/thaisrocafull/Documents/CURSOS_I_APUNTS/data_science/SPRINT 3


In [4]:
DF_ORIGIN = pd.read_csv ('DelayedFlights.csv') # creamos objeto con el nombre de la variable. Luego al hacer modificaciones , seguro que hago copia.
pd.set_option('display.max.columns', None)
DF_ORIGIN

Unnamed: 0  Year  Month  DayofMonth  DayOfWeek  DepTime  CRSDepTime  \
0                 0  2008      1           3          4   2003.0        1955   
1                 1  2008      1           3          4    754.0         735   
2                 2  2008      1           3          4    628.0         620   
3                 4  2008      1           3          4   1829.0        1755   
4                 5  2008      1           3          4   1940.0        1915   
...             ...   ...    ...         ...        ...      ...         ...   
1936753     7009710  2008     12          13          6   1250.0        1220   
1936754     7009717  2008     12          13          6    657.0         600   
1936755     7009718  2008     12          13          6   1007.0         847   
1936756     7009726  2008     12          13          6   1251.0        1240   
1936757     7009727  2008     12          13          6   1110.0        1103   

         ArrTime  CRSArrTime UniqueCarrier  FlightNum TailNum  \
0         2211.0        2225            WN        335  N712SW   
1         1002.0        1000            WN       3231  N772SW   
2          804.0         750            WN        448  N428WN   
3         1959.0        1925            WN       3920  N464WN   
4         2121.0        2110            WN        378  N726SW   
...          ...         ...           ...        ...     ...   
1936753   1617.0        1552            DL       1621  N938DL   
1936754    904.0         749            DL       1631  N3743H   
1936755   1149.0        1010            DL       1631  N909DA   
1936756   1446.0        1437            DL       1639  N646DL   
1936757   1413.0        1418            DL       1641  N908DL   

         ActualElapsedTime  CRSElapsedTime  AirTime  ArrDelay  DepDelay  \
0                    128.0           150.0    116.0     -14.0       8.0   
1                    128.0           145.0    113.0       2.0      19.0   
2                     96.0            90.0     76.0      14.0       8.0   
3                     90.0            90.0     77.0      34.0      34.0   
4                    101.0           115.0     87.0      11.0      25.0   
...                    ...             ...      ...       ...       ...   
1936753              147.0           152.0    120.0      25.0      30.0   
1936754              127.0           109.0     78.0      75.0      57.0   
1936755              162.0           143.0    122.0      99.0      80.0   
1936756              115.0           117.0     89.0       9.0      11.0   
1936757              123.0           135.0    104.0      -5.0       7.0   

        Origin Dest  Distance  TaxiIn  TaxiOut  Cancelled CancellationCode  \
0          IAD  TPA       810     4.0      8.0          0                N   
1          IAD  TPA       810     5.0     10.0          0                N   
2          IND  BWI       515     3.0     17.0          0                N   
3          IND  BWI       515     3.0     10.0          0                N   
4          IND  JAX       688     4.0     10.0          0                N   
...        ...  ...       ...     ...      ...        ...              ...   
1936753    MSP  ATL       906     9.0     18.0          0                N   
1936754    RIC  ATL       481    15.0     34.0          0                N   
1936755    ATL  IAH       689     8.0     32.0          0                N   
1936756    IAD  ATL       533    13.0     13.0          0                N   
1936757    SAT  ATL       874     8.0     11.0          0                N   

         Diverted  CarrierDelay  WeatherDelay  NASDelay  SecurityDelay  \
0               0           NaN           NaN       NaN            NaN   
1               0           NaN           NaN       NaN            NaN   
2               0           NaN           NaN       NaN            NaN   
3               0           2.0           0.0       0.0            0.0   
4               0           NaN           NaN       NaN    

In [5]:
DF_ORIGIN.info() # Primer vistazo del df 

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1936758 entries, 0 to 1936757
Data columns (total 30 columns):
 #   Column             Dtype  
---  ------             -----  
 0   Unnamed: 0         int64  
 1   Year               int64  
 2   Month              int64  
 3   DayofMonth         int64  
 4   DayOfWeek          int64  
 5   DepTime            float64
 6   CRSDepTime         int64  
 7   ArrTime            float64
 8   CRSArrTime         int64  
 9   UniqueCarrier      object 
 10  FlightNum          int64  
 11  TailNum            object 
 12  ActualElapsedTime  float64
 13  CRSElapsedTime     float64
 14  AirTime            float64
 15  ArrDelay           float64
 16  DepDelay           float64
 17  Origin             object 
 18  Dest               object 
 19  Distance           int64  
 20  TaxiIn             float64
 21  TaxiOut            float64
 22  Cancelled          int64  
 23  CancellationCode   object 
 24  Diverted           int64  
 25  CarrierDelay      

### Analisis del DF, a fin de entenderlo y detectar columnas que nos sirven. 

0. **Unnamed: 0: NO NOS SIRVE. trabajaremos con el índice creado por el programa.
1. *Year: El año del vuelo. COMO SOLO HAY UN AÑO , O LO AGRUPAMOS EN LA FECHA O PONEMOS TITULO DEL 2008 y eliminamos la columna.
2. Month: El mes del vuelo. (este dato y el siguiente, podrian tener formato fecha si quisieramos)
3. DayofMonth: El día del mes del vuelo.
4. DayOfWeek: El día de la semana del vuelo. *OPC1: LA USARÉ PARA SABER QUÉ DÍA DE LA SEMANA HAY MAS RETRASOS
5. DepTime: La hora de salida real en formato ***decimal.(PERO VEO QUE TODOS LOS DECIMALES SON 0 EN COLUMNAS- Entiendo que el formato es hhmm.0).
6. CRSDepTime: La hora de salida programada en formato **entero. (hora local.)
7. ArrTime: La hora de llegada real en formato ***decimal.  TENER EN CUENTA QUE ES HORA LOCAL DEL AEROPUERTO DE LLEGADA. Eso indica que no nos serviría a la hora de deducir datos de retraso (sin contar con una tabla de diferencia horarias por zona) Supongo que se podría eliminar pero quizás quedaría raro en la tabla.
8. CRSArrTime: La hora de llegada programada en formato **entero.
9. UniqueCarrier: El identificador único de la aerolínea. *OPC2: LA USARÉ PARA SABER AEROLINIA CON MAS RETRASOS P.EJ.
10. FlightNum: El número de vuelo. 
11. TailNum: El número de cola de la aeronave. **Identifica el Avión.
12. ActualElapsedTime: El tiempo de vuelo real en formato ***decimal.EN ESTE CASO INCLUYE ATERRIZAJE, ASÍ QUE PARA SABER LA VELOCIDAD NOS BASAREMOS EN EL AIRTIME DEL PUNTO 14.
13. CRSElapsedTime: El tiempo de vuelo programado en formato ***decimal. CREO QUE NO NOS HARÁ FALTA.
14. AirTime: El tiempo que la aeronave estuvo en el aire en formato ***decimal. TIEMPO DE VUELO, EN AIRE, (IMPORTANTE PARA DEDUCIR LA VELOCIDAD).
15. *ArrDelay: El retraso en la llegada en formato *decimal. *ES, EN RESUMEN, EL RETRASO TOTAL EN MINUTOS.
16. *DepDelay: El retraso en la salida en formato *decimal.
17. Origin: El código del aeropuerto de origen.
18. Dest: El código del aeropuerto de destino.
19. **Distance: La distancia del vuelo en millas. LA NECESITAMOS PARA CALCULAR LA VELOCIDAD
20. *TaxiIn: El tiempo que llevó el taxi después del aterrizaje en formato decimal. *VER EN QUÉ TIPO DE RETRASO ENCAJA.No sabemos si este retraso forma parte del tiempo de retraso de la compañía o otros.
21. *TaxiOut: El tiempo que llevó el taxi antes del despegue en formato decimal.
22. Cancelled: Indica si el vuelo fue cancelado **Boleans (1 si fue cancelado, 0 en caso contrario). 
23. CancellationCode: Un código que indica la razón de la cancelación (por ejemplo, clima). OPC quizás para saber motivo de la cancelación.
24. *Diverted: Indica si el vuelo fue desviado (1 si fue desviado, 0 en caso contrario). Creo que será una columna complicada puesto que no nos dará datos concisos.
25. *CarrierDelay: Retraso causado por la aerolínea (en minutos) en formato ***decimal. TIPOS DE RETRASO
26. *WeatherDelay: Retraso causado por condiciones climáticas (en minutos) en formato ***decimal.TIPOS DE RETRASO
27. *NASDelay: Retraso causado por el Sistema Nacional de Espacio Aéreo (en minutos) en formato ***decimal.TIPOS DE RETRASO
28. *SecurityDelay: Retraso causado por problemas de seguridad (en minutos) en formato ***decimal.TIPOS DE RETRASO
29. *LateAircraftDelay: Retraso causado por la llegada tardía de la aeronave (en minutos) en formato ***decimal. TIPOS DE RETRASO en relación al retraso causado por la mísma compañía.

In [8]:
pd.unique(DF_ORIGIN['Year']) # para saber que todos pasan en el año 2008. 

array([2008])

In [9]:
DF_ORIGIN['Origin'].value_counts() # vemos que hay aeropuertos de todo el mundo y no solo de EUA.

ATL    131613
ORD    125979
DFW     95414
DEN     74323
LAX     58772
        ...  
BJI         4
PIR         3
PUB         2
INL         1
TUP         1
Name: Origin, Length: 303, dtype: int64

#### Este conjunto de datos contiene información sobre vuelos realizados en el año 2008 por diferentes compañías aéreas y en más de 300 aeropuertos distintos. Dicha tabla, incluye detalles sobre sus horas de salida y llegada, motivos de retraso, etc... 

### Comprobar los valores faltantes del DF

In [10]:
# Per contabilitzar els valors nuls a cada columna del dataset
(DF_ORIGIN.isnull().sum())

Unnamed: 0                0
Year                      0
Month                     0
DayofMonth                0
DayOfWeek                 0
DepTime                   0
CRSDepTime                0
ArrTime                7110
CRSArrTime                0
UniqueCarrier             0
FlightNum                 0
TailNum                   5
ActualElapsedTime      8387
CRSElapsedTime          198
AirTime                8387
ArrDelay               8387
DepDelay                  0
Origin                    0
Dest                      0
Distance                  0
TaxiIn                 7110
TaxiOut                 455
Cancelled                 0
CancellationCode          0
Diverted                  0
CarrierDelay         689270
WeatherDelay         689270
NASDelay             689270
SecurityDelay        689270
LateAircraftDelay    689270
dtype: int64

- Lo que más me impacta es la simetría de valores nulos en las columnas del motivo de retraso. Seguramente tiene una relación con el número de vuelos que no se han retrasado (o retraso inferior a 15') o que han sido desviados o cancelados. 
- Los resultados también són iguales en ActualElapsedTime, CRSElapsedTime, AirTime,ArrDelay... por lo que entiendo que deriban de vuelos no realizados o algo así. 

Procedo a verificar mi hipótesis de lo citado en el párrafo anterior: relación de vuelos no ejecutadoso o sin retraso.

In [11]:
vuelos_sin_retraso = (DF_ORIGIN['ArrDelay'] < 15).value_counts()[True] # sumamos no retrasos . También se podia hacer con shape[0]
vuelos_sin_retraso

680883

In [12]:
vuelos_derivados = (DF_ORIGIN['Diverted']== 1).value_counts() # sumamos vuelos desviados
vuelos_derivados

False    1929004
True        7754
Name: Diverted, dtype: int64

In [13]:
vuelos_cancelados = (DF_ORIGIN['Cancelled']==1).value_counts() # sumamos los vuelos cancelados
vuelos_cancelados

False    1936125
True         633
Name: Cancelled, dtype: int64

In [14]:
suma_valores_nan= vuelos_sin_retraso + vuelos_derivados+ vuelos_cancelados #también SUM_CANCEL = np.sum([, ,])
suma_valores_nan

False    4546012
True      689270
dtype: int64

- Con esta suma de las tres variables, podemos comprobar que los datos NaN de las columnas que justifican el motivo de retraso, son igual a 0. Porque efectivamente no lo ha habido. Procederemos a sustituirlos por 0.

Decido hacer una nueva tabla y,
- primero, retirando las filas de los vuelos cancelados y derivados, pues nos desviarían los resultados de los retrasos. Equivale a las columnas cuando ['Cancelled']==1 y ['Diverted']== 1.
- Segundo, Sustituyendo los valores NaN por 0.

También observamos que los valores NaN de la columna:  ActualElapsedTime, AirTime y ArrDelay que equivalen a 8387, son la suma de los vuelos cancelados y desviados. Por lo tanto los datos también pueden ser 0. Aunque quizás al corregir los otros, se corrigen solos.

In [15]:
DF_net = DF_ORIGIN[(DF_ORIGIN['Cancelled'] == False) & (DF_ORIGIN['Diverted'] == False)]

In [16]:
DF_net['Cancelled'].value_counts()

0    1928371
Name: Cancelled, dtype: int64

In [17]:
DF_net['Diverted'].value_counts()

0    1928371
Name: Diverted, dtype: int64

In [18]:
col_rempl = ['CarrierDelay', 'WeatherDelay', 'NASDelay', 'SecurityDelay', 'LateAircraftDelay','TailNum',] #también aprovecho para cambiar los nans a 0 de TailNum.

In [19]:
DF_net.loc[:, col_rempl] = DF_net[col_rempl].fillna(0)  #DF_net[col_rempl] = DF_net[col_rempl].fillna(0) otra manera de hacerlo pero debo acabar de entender porque uno cambia el df original y el otro no. 

/var/folders/b5/qd0s7b0d7cg48d0n9kg8c6lr0000gn/T/ipykernel_68516/706567964.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  DF_net.loc[:, col_rempl] = DF_net[col_rempl].fillna(0)  #DF_net[col_rempl] = DF_net[col_rempl].fillna(0) otra manera de hacerlo pero debo acabar de entender porque uno cambia el df original y el otro no.


In [20]:
(DF_net.isnull().sum())

Unnamed: 0           0
Year                 0
Month                0
DayofMonth           0
DayOfWeek            0
DepTime              0
CRSDepTime           0
ArrTime              0
CRSArrTime           0
UniqueCarrier        0
FlightNum            0
TailNum              0
ActualElapsedTime    0
CRSElapsedTime       0
AirTime              0
ArrDelay             0
DepDelay             0
Origin               0
Dest                 0
Distance             0
TaxiIn               0
TaxiOut              0
Cancelled            0
CancellationCode     0
Diverted             0
CarrierDelay         0
WeatherDelay         0
NASDelay             0
SecurityDelay        0
LateAircraftDelay    0
dtype: int64

- Vemos como desaparecen los valores NaN derivados de columnas y filas que hemos eliminado, por ejemplo los Taxi out y taxi in, pues seguro que no se utilizaron si hubieron cancelaciones.pueden ser que no se haya usado por cercanía del avión a la puerta de desembarque.

Ahora voy a borrar las columnas que no me sirven: Unnamed: 0, Diverted, Cancelled, CancellationCode, TaxiIN, TaxiOut.

In [21]:
DF_net = DF_net.drop(columns= ['Unnamed: 0', 'Diverted', 'Cancelled', 'CancellationCode', 'TaxiIn','TaxiOut'],axis=1)

In [22]:
print(DF_net.columns)

Index(['Year', 'Month', 'DayofMonth', 'DayOfWeek', 'DepTime', 'CRSDepTime',
       'ArrTime', 'CRSArrTime', 'UniqueCarrier', 'FlightNum', 'TailNum',
       'ActualElapsedTime', 'CRSElapsedTime', 'AirTime', 'ArrDelay',
       'DepDelay', 'Origin', 'Dest', 'Distance', 'CarrierDelay',
       'WeatherDelay', 'NASDelay', 'SecurityDelay', 'LateAircraftDelay'],
      dtype='object')


### 1.2- Reducir la dimensión aleatoriamente a 200,000 registros

In [23]:
DF_RED= DF_net.sample(n=200000, random_state=1)
pd.set_option('display.expand_frame_repr', False)
DF_RED

Year  Month  DayofMonth  DayOfWeek  DepTime  CRSDepTime  ArrTime  CRSArrTime UniqueCarrier  FlightNum TailNum  ActualElapsedTime  CRSElapsedTime  AirTime  ArrDelay  DepDelay Origin Dest  Distance  CarrierDelay  WeatherDelay  NASDelay  SecurityDelay  LateAircraftDelay
946177   2008      6           7          6   1121.0        1042   1257.0        1214            OO       6452  N727SK               96.0            92.0     73.0      43.0      39.0    TUS  LAX       451          39.0           0.0       4.0            0.0                0.0
1490024  2008      9          22          1   1844.0        1835   1946.0        1945            MQ       3428  N612AE               62.0            70.0     50.0       1.0       9.0    MAF  DFW       309           0.0           0.0       0.0            0.0                0.0
1572676  2008     10          18          6   1624.0        1615   1716.0        1717            US       1228  N506AU              112.0           122.0     91.0      -1.0       9.0    CLT  ORD       599           0.0           0.0       0.0            0.0                0.0
1261096  2008      7           8          2   1748.0        1720   2017.0        1920            DL        822  N907DE              149.0           120.0     85.0      57.0      28.0    ATL  DCA       547           0.0           0.0      57.0            0.0                0.0
653536   2008      4           4          5   1653.0        1640   1725.0        1717            EV       4158  N839AS               92.0            97.0     77.0       8.0      13.0    CVG  MEM       403           0.0           0.0       0.0            0.0                0.0
...       ...    ...         ...        ...      ...         ...      ...         ...           ...        ...     ...                ...             ...      ...       ...       ...    ...  ...       ...           ...           ...       ...            ...                ...
134447   2008      1           7          1    211.0         100    702.0         556            NW        774  N505US              171.0           176.0    144.0      66.0      71.0    LAS  MSP      1300          59.0           0.0       0.0            0.0                7.0
1020385  2008      6           6          5   1806.0        1800   1952.0        1945            MQ       4664  N700LE              106.0           105.0     79.0       7.0       6.0    BDL  RDU       532           0.0           0.0       0.0            0.0                0.0
1667035  2008     11          15          6   1702.0        1655   1747.0        1808            UA        411  N371UA              105.0           133.0     89.0     -21.0       7.0    CLT  ORD       599           0.0           0.0       0.0            0.0                0.0
623129   2008      4          27          7   1301.0        1255   1428.0        1429            OO       5866  N926SW               87.0            94.0     70.0      -1.0       6.0    ORD  LNK       466           0.0           0.0       0.0            0.0                0.0
120102   2008      1           6          7   1926.0        1920   2242.0        2240            MQ       3725  N933JN              136.0           140.0    118.0       2.0       6.0    DFW  CAE       922           0.0           0.0       0.0            0.0                0.0

[200000 rows x 24 columns]

In [24]:
DF_RED.shape

(200000, 24)

In [25]:
DF_RED.describe().astype(int).T #esta T intercambia filas / columnas para ser mas visual. 

count  mean   std   min   25%   50%   75%   max
Year               200000  2008     0  2008  2008  2008  2008  2008
Month              200000     6     3     1     3     6     9    12
DayofMonth         200000    15     8     1     8    16    23    31
DayOfWeek          200000     3     1     1     2     4     6     7
DepTime            200000  1518   451     1  1203  1545  1900  2400
CRSDepTime         200000  1468   424     0  1135  1510  1815  2359
ArrTime            200000  1610   548     1  1316  1715  2030  2400
CRSArrTime         200000  1634   464     0  1325  1706  2014  2359
FlightNum          200000  2183  1948     1   608  1537  3427  9741
ActualElapsedTime  200000   133    72    16    80   116   165   727
CRSElapsedTime     200000   134    71    15    81   116   165   660
AirTime            200000   108    68     1    58    90   137   641
ArrDelay           200000    42    57   -55     9    25    56  2461
DepDelay           200000    43    53     6    12    24    54  2467
Distance           200000   764   574    31   337   606   997  4962
CarrierDelay       200000    12    36     0     0     0    10  2436
WeatherDelay       200000     2    16     0     0     0     0   902
NASDelay           200000     9    28     0     0     0     6   992
SecurityDelay      200000     0     1     0     0     0     0   211
LateAircraftDelay  200000    16    36     0     0     0    18  1002

- Quiero ver en detalle el vuelo donde AirTime es de 1'

In [148]:
indice_min_airtime = DF_RED['AirTime'].idxmin()
vuelo_min_airtime = DF_RED.loc[indice_min_airtime]
print(vuelo_min_airtime) 

Year                   2008
Month                     4
DayofMonth                4
DayOfWeek                 5
DepTime              1945.0
CRSDepTime             1935
ArrTime              2105.0
CRSArrTime             2110
UniqueCarrier            OH
FlightNum              5100
TailNum              N708CA
ActualElapsedTime      80.0
CRSElapsedTime         95.0
AirTime                 1.0
ArrDelay               -5.0
DepDelay               10.0
Origin                  CVG
Dest                    ROC
Distance                461
CarrierDelay            0.0
WeatherDelay            0.0
NASDelay                0.0
SecurityDelay           0.0
LateAircraftDelay       0.0
Delay                    No
Vel(mph)              27660
Name: 615069, dtype: object


### Exercici 2 Fes un informe complet del dataset:
Crea columnes noves (velocitat mitjana del vol, si ha arribat tard o no...). 

In [142]:
# creo el DF para volcar los resultados Y/N de si el vuelo lleva retraso o no
col_DF_retard = (DF_RED['ArrDelay']>=15).apply(lambda x: 'Yes' if x else 'No')
col_DF_retard.rename('Delay', inplace=True)

946177     Yes
1490024     No
1572676     No
1261096    Yes
653536      No
          ... 
134447     Yes
1020385     No
1667035     No
623129      No
120102      No
Name: Delay, Length: 200000, dtype: object

In [30]:
col_DF_retard.value_counts()

Yes    129578
No      70422
Name: RET Y/N, dtype: int64

In [33]:
DF_RED['Delay']= col_DF_retard 
DF_RED

Year  Month  DayofMonth  DayOfWeek  DepTime  CRSDepTime  ArrTime  CRSArrTime UniqueCarrier  FlightNum TailNum  ActualElapsedTime  CRSElapsedTime  AirTime  ArrDelay  DepDelay Origin Dest  Distance  CarrierDelay  WeatherDelay  NASDelay  SecurityDelay  LateAircraftDelay Delay
946177   2008      6           7          6   1121.0        1042   1257.0        1214            OO       6452  N727SK               96.0            92.0     73.0      43.0      39.0    TUS  LAX       451          39.0           0.0       4.0            0.0                0.0   Yes
1490024  2008      9          22          1   1844.0        1835   1946.0        1945            MQ       3428  N612AE               62.0            70.0     50.0       1.0       9.0    MAF  DFW       309           0.0           0.0       0.0            0.0                0.0    No
1572676  2008     10          18          6   1624.0        1615   1716.0        1717            US       1228  N506AU              112.0           122.0     91.0      -1.0       9.0    CLT  ORD       599           0.0           0.0       0.0            0.0                0.0    No
1261096  2008      7           8          2   1748.0        1720   2017.0        1920            DL        822  N907DE              149.0           120.0     85.0      57.0      28.0    ATL  DCA       547           0.0           0.0      57.0            0.0                0.0   Yes
653536   2008      4           4          5   1653.0        1640   1725.0        1717            EV       4158  N839AS               92.0            97.0     77.0       8.0      13.0    CVG  MEM       403           0.0           0.0       0.0            0.0                0.0    No
...       ...    ...         ...        ...      ...         ...      ...         ...           ...        ...     ...                ...             ...      ...       ...       ...    ...  ...       ...           ...           ...       ...            ...                ...   ...
134447   2008      1           7          1    211.0         100    702.0         556            NW        774  N505US              171.0           176.0    144.0      66.0      71.0    LAS  MSP      1300          59.0           0.0       0.0            0.0                7.0   Yes
1020385  2008      6           6          5   1806.0        1800   1952.0        1945            MQ       4664  N700LE              106.0           105.0     79.0       7.0       6.0    BDL  RDU       532           0.0           0.0       0.0            0.0                0.0    No
1667035  2008     11          15          6   1702.0        1655   1747.0        1808            UA        411  N371UA              105.0           133.0     89.0     -21.0       7.0    CLT  ORD       599           0.0           0.0       0.0            0.0                0.0    No
623129   2008      4          27          7   1301.0        1255   1428.0        1429            OO       5866  N926SW               87.0            94.0     70.0      -1.0       6.0    ORD  LNK       466           0.0           0.0       0.0            0.0                0.0    No
120102   2008      1           6          7   1926.0        1920   2242.0        2240            MQ       3725  N933JN              136.0           140.0    118.0       2.0       6.0    DFW  CAE       922           0.0           0.0       0.0            0.0                0.0    No

[200000 rows x 25 columns]

- En esta tabla hemos añadido la columna Delay que nos muestra en forma booleana Y/N si el vuelo ha tenido retraso.

In [38]:
# Caluculo de la velocidad mediana de vuelo (distancia/tiempo en minutos)
DF_RED['Vel(mph)'] = DF_RED['Distance'] / (DF_RED['AirTime'] / 60)
DF_RED['Vel(mph)'] = DF_RED['Vel(mph)'].astype(int)
DF_RED[['FlightNum', 'Distance', 'AirTime', 'Vel(mph)']]

FlightNum  Distance  AirTime  Vel(mph)
946177        6452       451     73.0       370
1490024       3428       309     50.0       370
1572676       1228       599     91.0       394
1261096        822       547     85.0       386
653536        4158       403     77.0       314
...            ...       ...      ...       ...
134447         774      1300    144.0       541
1020385       4664       532     79.0       404
1667035        411       599     89.0       403
623129        5866       466     70.0       399
120102        3725       922    118.0       468

[200000 rows x 4 columns]

In [143]:
DF_RED.head()

Year  Month  DayofMonth  DayOfWeek  DepTime  CRSDepTime  ArrTime  CRSArrTime UniqueCarrier  FlightNum TailNum  ActualElapsedTime  CRSElapsedTime  AirTime  ArrDelay  DepDelay Origin Dest  Distance  CarrierDelay  WeatherDelay  NASDelay  SecurityDelay  LateAircraftDelay Delay  Vel(mph)
946177   2008      6           7          6   1121.0        1042   1257.0        1214            OO       6452  N727SK               96.0            92.0     73.0      43.0      39.0    TUS  LAX       451          39.0           0.0       4.0            0.0                0.0   Yes       370
1490024  2008      9          22          1   1844.0        1835   1946.0        1945            MQ       3428  N612AE               62.0            70.0     50.0       1.0       9.0    MAF  DFW       309           0.0           0.0       0.0            0.0                0.0    No       370
1572676  2008     10          18          6   1624.0        1615   1716.0        1717            US       1228  N506AU              112.0           122.0     91.0      -1.0       9.0    CLT  ORD       599           0.0           0.0       0.0            0.0                0.0    No       394
1261096  2008      7           8          2   1748.0        1720   2017.0        1920            DL        822  N907DE              149.0           120.0     85.0      57.0      28.0    ATL  DCA       547           0.0           0.0      57.0            0.0                0.0   Yes       386
653536   2008      4           4          5   1653.0        1640   1725.0        1717            EV       4158  N839AS               92.0            97.0     77.0       8.0      13.0    CVG  MEM       403           0.0           0.0       0.0            0.0                0.0    No       314

- Hemos añadido la columna de velocidad media a la tabla

### Fes una taula de les aerolínies amb més endarreriments acumulats. 

In [141]:
# Creamos una tabla donde constaran los vuelos agrupados por compañía aerea 'UniqueCarrier' donde consten la cantidad de vuelos atrasados  >=15 * podriamos usar la columna creada anteriormente Delay y True, 'ArrDelay'. y la cantidad de MINUTOS de retraso acomulado
vols_retard_15 = DF_RED[DF_RED['ArrDelay'] >= 15]
AER_ATR = vols_retard_15.groupby('UniqueCarrier').agg(Q_Vuelos_Atrasados=('ArrDelay', 'count'), Sum_Minutos_de_Retraso=('ArrDelay', 'sum')).sort_values(by= 'Sum_Minutos_de_Retraso',ascending=False)
AER_ATR.head()

Q_Vuelos_Atrasados  Sum_Minutos_de_Retraso
UniqueCarrier                                            
WN                          21048               1116046.0
AA                          13785                913330.0
UA                           9992                697799.0
MQ                          10209                660352.0
OO                           9074                586574.0

- Aqui nos muestra los 5 primeros resultados de la tabla ordenada por el máximo de minutos de retraso. Que en las dos primeras compañías, también coincide con el número de vuelos atrasados.

### Tabla con las rutas más largas

In [106]:
# Necesito las columnas : FlightNum, Origin, Dest, Distance
RETR_RUTAS = vols_retard_15[['FlightNum', 'Origin', 'Dest', 'Distance', 'ArrDelay']]

# Reemplaza valores negativos en la columna 'ArrDelay' por 0 para que no me distorsione los valores.
RETR_RUTAS['ArrDelay'] = RETR_RUTAS['ArrDelay'].apply(lambda x: max(0, x))

# Agrupa por 'FlightNum', 'Origin' y 'Dest' y calcula la suma de minutos de retraso
RETR_RUTAS = RETR_RUTAS.groupby(['FlightNum', 'Origin', 'Dest']).agg({'Distance': 'max', 'ArrDelay': 'sum'}).reset_index().sort_values(by='Distance', ascending=False)

RUTAS_UNICAS = RETR_RUTAS.drop_duplicates(subset=['Origin', 'Dest'])
RUTAS_UNICAS.head(10)


/var/folders/b5/qd0s7b0d7cg48d0n9kg8c6lr0000gn/T/ipykernel_68516/2592715803.py:5: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  RETR_RUTAS['ArrDelay'] = RETR_RUTAS['ArrDelay'].apply(lambda x: max(0, x))


FlightNum Origin Dest  Distance  ArrDelay
213           15    EWR  HNL      4962     698.0
195           14    HNL  EWR      4962     115.0
17313       1273    ATL  HNL      4502     427.0
17422       1282    HNL  ATL      4502     171.0
1182          73    ORD  HNL      4243     382.0
1164          72    HNL  ORD      4243     158.0
34             3    KOA  ORD      4213     112.0
14038       1003    ORD  OGG      4184      17.0
11913        809    MSP  HNL      3972     203.0
9409         624    HNL  MSP      3972      45.0

- Contemplando la tabla y teniendo en cuenta los vuelos de ida y vuelta. Las 6 rutas más largas en términos de distancia serían las de : EWR/HNL, ATL/HNL, ORD/HNL, KOA/ORD, ORD/OGG, MSP/HNL

In [145]:
max_distance = DF_RED['Distance'].max()
print("La ruta con máxima distancia es la ofrecida por los vuelos números 14 y 15, de HNL a EWR. \nY equivale a",max_distance,"millas")

La ruta con máxima distancia es la ofrecida por los vuelos números 14 y 15, de HNL a EWR. 
Y equivale a 4962 millas


In [146]:
# Filtra las filas donde 'ActualElapsedTime' es igual al tiempo máximo
vuelo_mas_largo = DF_RED[DF_RED['ActualElapsedTime'] == max_air]

# Selecciona las columnas 'FlightNum', 'Origin', 'Dest' y 'ActualElapsedTime'
vuelo_mas_largo = vuelo_mas_largo[['FlightNum', 'Origin', 'Dest', 'ActualElapsedTime']]
print(vuelo_mas_largo)

        FlightNum Origin Dest  ActualElapsedTime
528094          5    DFW  HNL              727.0


- El vuelo más largo en tiempo es el número 5 que traza la ruta DFW/HNL y tiene una duración de 727 minutos (algo más de 12 horas de vuelo).

### Tabla con las columnas que acumulan más atrasos

In [147]:
# Ordena por la columna 'ArrDelay' (suma de minutos de retraso) en orden descendente
RUTAS_MAS_ATRASO = RETR_RUTAS.sort_values(by='ArrDelay', ascending=False)
top_5_rutas_atraso = RUTAS_MAS_ATRASO.head(5)
top_5_rutas_atraso

FlightNum Origin Dest  Distance  ArrDelay
30999       2994    CLT  EWR       529    3102.0
8002         527    EWR  MCO       938    2925.0
26817       2362    EWR  RIC       278    2701.0
21997       1699    CLT  MSP       930    2673.0
20209       1555    MIA  SJU      1045    2659.0

- Esta tabla nos muestra los vuelos con más retrasos, ordenados de manera descendiente.

### Exercici 3 Exporta el dataset net i amb les noves columnes a Excel.

In [65]:
# Exporto el DF amb les dos noves columnes a excel
DF_RED.to_excel('DelayedFlights_tha.xlsx')